android/app/src/main/java/com/now/naaga/presentation/custom/GameButton.kt

```diff
@@ -0,0 +1,126 @@
+import android.graphics.Color
+import android.graphics.Paint
+import android.graphics.RectF
+import android.util.AttributeSet
+import android.view.MotionEvent
+import androidx.appcompat.widget.AppCompatButton
+import com.now.naaga.R
+
+class GameButton(context: Context, attrs: AttributeSet? = null) : AppCompatButton(context, attrs) {
+    private val radius: Float
+    private var clicked: Boolean = false
+    private var clickAction: OnClickListener? = null
+
+    private val mainColor: Int
+    private val firstShadowColor: Int
+    private val middleColor: Int
+    private val secondShadowColor: Int
+    private val bottomColor: Int
+
+    init {
+        context.theme.obtainStyledAttributes(
+            attrs,
+            R.styleable.GameButton,
+            0,
+            0,
+        ).apply {
+            radius = getDimensionPixelSize(R.styleable.GameButton_radius, 0).toFloat()
+            val gameButtonColor = GameButtonColor.getColor((getInteger(R.styleable.GameButton_buttonColor, 0)))
+            mainColor = Color.parseColor(gameButtonColor.mainColor)
+            firstShadowColor = Color.parseColor(gameButtonColor.firstShadowColor)
+            middleColor = Color.parseColor(gameButtonColor.middleColor)
+            secondShadowColor = Color.parseColor(gameButtonColor.secondShadowColor)
+            bottomColor = Color.parseColor(gameButtonColor.bottomColor)
+            recycle()
+        }
+    }
+
+    private fun getPaint(color: Int) = Paint().apply {
+        this.color = color
+    }
+
+    private val ripplePaint = Paint().apply {
+        this.color = Color.parseColor("#4D000000")
+    }
+
+    override fun onDraw(canvas: Canvas) {
+        setBackgroundColor(Color.TRANSPARENT)
+        drawButton(canvas)
+        if (clicked) drawRipple(canvas)
+        super.onDraw(canvas)
+    }
+
+    override fun onTouchEvent(event: MotionEvent): Boolean {
+        when (event.action) {
+            MotionEvent.ACTION_DOWN -> {
+                clicked = true
+                invalidate()
+            }
+
+            MotionEvent.ACTION_UP -> {
+                clicked = false
+                invalidate()
+                clickAction?.onClick(this)
+            }
+        }
+        return true
+    }
+
+    override fun setOnClickListener(l: OnClickListener?) {
+        clickAction = l
+    }
+
+    private fun drawRipple(canvas: Canvas) {
+        canvas.drawRoundRect(getBottomRect(), radius, radius, ripplePaint)
+    }
+
+    private fun drawButton(canvas: Canvas) {
+        canvas.drawRoundRect(getBottomRect(), radius, radius, getPaint(bottomColor))
+        canvas.drawRoundRect(getSecondShadowRect(), radius, radius, getPaint(secondShadowColor))
+        canvas.drawRoundRect(getMiddleRect(), radius, radius, getPaint(middleColor))
+        canvas.drawRoundRect(getFirstShadowRect(), radius, radius, getPaint(firstShadowColor))
+        canvas.drawRoundRect(getMainRect(), radius, radius, getPaint(mainColor))
```

Comment:
[P5]
```suggestion
        with(canvas) {
            drawRoundRect(getBottomRect(), radius, radius, getPaint(bottomColor))
            drawRoundRect(getSecondShadowRect(), radius, radius, getPaint(secondShadowColor))
            drawRoundRect(getMiddleRect(), radius, radius, getPaint(middleColor))
            drawRoundRect(getFirstShadowRect(), radius, radius, getPaint(firstShadowColor))
            drawRoundRect(getMainRect(), radius, radius, getPaint(mainColor))
        }
```
with를 사용해서 반복을 줄일 수 있을 것 같아요!

Reply:
반영했습니다.